test/schema.uuid.test.js

```diff
@@ -0,0 +1,56 @@
+'use strict';
+
+const start = require('./common');
+const util = require('./util');
+
+const assert = require('assert');
+
+const mongoose = start.mongoose;
+const Schema = mongoose.Schema;
+
+describe('SchemaUUID', function() {
+  let Model;
+  let TestSchema;
+  let db;
+
+  before(async function() {
+    TestSchema = new Schema({ x: { type: mongoose.Schema.Types.UUID } });
+    db = await start().asPromise();
+  });
+
+  after(async function() {
+    await db.close();
+  });
+
+  beforeEach(async() => {
+    await db.deleteModel(/.*/);
+    Model = db.model('Test', TestSchema);
+  });
+  afterEach(() => util.clearTestData(db));
+  afterEach(() => util.stopRemainingOps(db));
+
+  it('basic functionality should work', async function() {
+    const doc = new Model({ x: '09190f70-3d30-11e5-8814-0f4df9a59c41' });
+    assert.ifError(doc.validateSync());
+    assert.ok(typeof doc.x === 'string');
+    assert.strictEqual(doc.x, '09190f70-3d30-11e5-8814-0f4df9a59c41');
+    await doc.save();
+
+    const query = Model.findOne({ x: '09190f70-3d30-11e5-8814-0f4df9a59c41' });
```

Comment:
I'd like to see a test using `$lte` and `$nin`

Reply:
will try to add one

Reply:
i just looked into `$lte`, though i dont quite understand how that is supposed to work with UUID, because from what i can tell there is not "less than" or "higher than" in UUID (at least in the generic implementation and not having a specific version)

also mongodb does not list anything that `$lte` (or `$lt`, `$gt`, `$gte`) would work on UUID, do they work on buffers?

Reply:
@hasezoey 
Assuming that we store UUIDs as strings in MongoDB, $lt/$gt in mongo generally works with strings by sorting alphabetically. Check this [mongo playground](https://mongoplayground.net/p/UjrfQpq9mqa), for example.

I fail to see how that can be useful for UUIDs, though. AFAIK, UUIDs don't have any guarantees for alphabetical sorting.

Reply:
> Assuming that we store UUIDs as strings in MongoDB

they are not, we store them as BSON UUID Buffers (bin-type 4)

Reply:
Interesting. I don't know what the difference from a developer experience would be like. Care to explain to me what some pros and cons of storing them as UUID buffers instead of strings are?

Reply:
> Interesting. I don't know what the difference from a developer experience would be like. Care to explain to me what some pros and cons of storing them as UUID buffers instead of strings are?

i dont know, but mongodb supports this Buffer type and so is probably more efficient in storing that, and using "it raw" (with the mongo client instead of mongoose), it is kinda cumbersome to use it this way, but with mongoose we can provide a easy back-end by converting it to string (like the current implementation) and the users using mongoose dont have to worry about the back-end buffer

Reply:
It is a decision made by mongodb that bson stores uuids as buffers. The same as ObjectIds are stored as buffers. Only issue is that uuid version 1 to 5 are not incrementing sequentially. But there is the proposal  of uuid v6, which should do increment sequentially. 

http://gh.peabody.io/uuidv6/

Thats why I like that we can provide a uuid instead of forcing the use of v1 or v4. So if somebody wants to use v6, it can be done.

Reply:
> instead of forcing the use of v1 or v4.

we currently dont force a type, mongodb allows version 0-9 (read like a regex), and what i discussed earlier was custom validators on mongoose's side

Reply:
tests for `$in` and `$nin` added in fd17d9bb26d5124c288098a24f0eaed647ffc09e

though i still dont know about `$lt` (and derivatives)

Reply:
Does MongoDB support `$gt`, `$lt`, etc. for buffers? There are times when I want to sort by `_id` in tests to ensure consistent sort order, regardless of what that sort order is. So it would be nice to support `$gt` and `$lt`, but not strictly necessary. If MongoDB doesn't support those with buffers, let's remove support for `$gt` and `$lt`.

I think it is fine to just support uuidv4 for now. If we want to support different uuid versions, that makes me more inclined to make uuid support a plugin.

Reply:
There is a plugin for mongoose 6 and uuid. But uuid is a native bson type. If it was some non-bson type then i would expect to be a plugin

Reply:
> Does MongoDB support $gt, $lt, etc. for buffers?

i dont quite know, but it seems at least mongoose lists it as supported for normal buffers

> But uuid is a native bson type. If it was some non-bson type then i would expect to be a plugin

that is also my reasoning

> I think it is fine to just support uuidv4 for now

currently we dont check for a specific version, though i can add it as a optional option like `enforceVersion: 4` (or something like that)

Reply:
We could just set the default field with crypto.randomUUID from node. If you want to have a different uuid version a implementer can just pass something like 

```js
const { v4, v1 } = require('uuid')

new Schema({ 
  x: { type: mongoose.Schema.Types.UUID, default: v4 },
  y: [{ type: mongoose.Schema.Types.UUID, default: v1 }]
})
```

Reply:
> We could just set the default field with crypto.randomUUID from node

would it be a good idea to provide a default for a specific version? when i should add it, how is this done in mongoose? (it probably is not as simple as setting `.defaultOptions = { default: generateV4 }`, is `.auto` the correct function?)